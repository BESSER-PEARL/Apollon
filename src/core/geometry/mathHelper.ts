export function isAlmostZero(value: number) {
    return Math.abs(value) < 1e-6;
}

export function areAlmostEqual(a: number, b: number) {
    return isAlmostZero(a - b);
}
