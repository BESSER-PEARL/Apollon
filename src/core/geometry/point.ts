import { isAlmostZero } from "./math/floatingPoint";

export interface Point {
    x: number;
    y: number;
}

export function pointsAreEqual(p: Point, q: Point) {
    const dx = Math.abs(p.x - q.x);
    const dy = Math.abs(p.y - q.y);

    return isAlmostZero(dx) && isAlmostZero(dy);
}

export function snapToGrid(position: number, gridSize: number): number {
    return Math.round(position / gridSize) * gridSize;
}

export function snapPointToGrid(p: Point, gridSize: number): Point {
    return {
        x: snapToGrid(p.x, gridSize),
        y: snapToGrid(p.y, gridSize)
    };
}
