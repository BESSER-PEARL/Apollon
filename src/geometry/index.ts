export * from "./delta";
export * from "./intersection";
export * from "./path";
export * from "./point";
export * from "./rect";
export * from "./size";
