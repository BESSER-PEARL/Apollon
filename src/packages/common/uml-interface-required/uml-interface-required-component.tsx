import React, { SFC } from 'react';
import { connect } from 'react-redux';
import { ModelState } from '../../../components/store/model-state';
import { UMLInterfaceRequired } from './uml-interface-required';
import { getOppositeDirection } from '../../../services/uml-element/uml-element-port';

const SIZE = 26;

type OwnProps = {
  element: UMLInterfaceRequired;
};
type StateProps = {
  hasOppositeRequiredInterface: boolean;
};
type DispatchProps = {};

type Props = OwnProps & StateProps & DispatchProps;

const enhance = connect<StateProps, DispatchProps, OwnProps, ModelState>((state, props) => {
  // filter all UMLInterfaceRequired
  const requiredInterfaces: UMLInterfaceRequired[] = state.diagram.ownedRelationships
    .map((relationshipId) => state.elements[relationshipId])
    .filter((relationship) => UMLInterfaceRequired.isUMLInterfaceRequired(relationship))
    .map((relationship) => relationship as UMLInterfaceRequired);

  // check if any other UMLInterfaceRequired has the same target as this element and if the direction of the UMLInterfaceRequired is the opposite
  return {
    hasOppositeRequiredInterface: requiredInterfaces
      .filter((element) => element.id !== props.element.id)
      .some(
        (otherRequiredInterface) =>
          otherRequiredInterface.target.element === props.element.target.element &&
          otherRequiredInterface.target.direction.valueOf() ===
            getOppositeDirection(props.element.target.direction).valueOf(),
      ),
  };
}, {});

const UMLInterfaceRequiredC: SFC<Props> = (props: Props) => {
  const { element, hasOppositeRequiredInterface } = props;
  return (
    <g>
      <marker
        id={`marker-${element.id}`}
        viewBox={`0 0 ${SIZE} ${SIZE}`}
        markerWidth={SIZE}
        markerHeight={SIZE}
        refX="0"
        refY="0"
        orient="auto"
        markerUnits="strokeWidth"
        strokeDasharray="1,0"
      >
        {/*M -> Move to, A -> Arc radiusX, radiusY, x-axis-rotation, bow-flag, endpointX,endpointY */}
        <path
          d={`M ${SIZE / 2 - (hasOppositeRequiredInterface ? 9 : 3)} -${
            (SIZE - (hasOppositeRequiredInterface ? 2 : 0)) / 2
          } a ${SIZE / 2},${SIZE / 2} 0 0 0 0,${SIZE - (hasOppositeRequiredInterface ? 2 : 0)}`}
          fill="none"
          stroke="black"
          strokeWidth={2}
        />
      </marker>
      <polyline
        points={element.path.map((point) => `${point.x} ${point.y}`).join(',')}
        stroke="black"
        fill="none"
        strokeWidth={1}
        markerEnd={`url(#marker-${element.id})`}
      />
    </g>
  );
};

export const UMLInterfaceRequiredComponent = enhance(UMLInterfaceRequiredC);
