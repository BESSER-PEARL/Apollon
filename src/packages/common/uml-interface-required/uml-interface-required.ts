import { UMLRelationship } from '../../../services/uml-relationship/uml-relationship';

export abstract class UMLInterfaceRequired extends UMLRelationship {
  static features = { ...UMLRelationship.features, variable: false };
}
