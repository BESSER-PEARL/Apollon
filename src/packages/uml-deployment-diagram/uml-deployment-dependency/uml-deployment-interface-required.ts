import { DeploymentRelationshipType } from '..';
import { UMLRelationship } from '../../../services/uml-relationship/uml-relationship';

export class UMLDeploymentDependency extends UMLRelationship {
  type = DeploymentRelationshipType.DeploymentDependency;
}
