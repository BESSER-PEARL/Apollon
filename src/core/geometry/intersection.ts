import { isAlmostZero } from "./math/floatingPoint";
import { Point } from "./point";
import {
    getBottomLeftCorner,
    getBottomRightCorner,
    getTopLeftCorner,
    getTopRightCorner,
    Rect
} from "./rect";

export function lineSegmentIntersectsRect(p: Point, q: Point, rect: Rect) {
    if (lineSegmentLiesWithinRect(p, q, rect)) {
        return true;
    }

    const topLeftCorner = getTopLeftCorner(rect);
    const topRightCorner = getTopRightCorner(rect);
    const bottomLeftCorner = getBottomLeftCorner(rect);
    const bottomRightCorner = getBottomRightCorner(rect);

    return (
        lineSegmentsIntersect(p, q, topLeftCorner, topRightCorner) ||
        lineSegmentsIntersect(p, q, topRightCorner, bottomRightCorner) ||
        lineSegmentsIntersect(p, q, topLeftCorner, bottomLeftCorner) ||
        lineSegmentsIntersect(p, q, bottomLeftCorner, bottomRightCorner)
    );
}

/**
 * Determines whether the given line lies entirely within the given rectangle.
 */
function lineSegmentLiesWithinRect(p: Point, q: Point, rect: Rect) {
    return p.x > rect.x && p.x < rect.x + rect.height && p.y > rect.y && p.y < rect.y + rect.height;
}

// Adapted from http://www.geeksforgeeks.org/check-if-two-given-line-segments-intersect/
function lineSegmentsIntersect(p1: Point, q1: Point, p2: Point, q2: Point) {
    const o1 = getOrientation(p1, q1, p2);
    const o2 = getOrientation(p1, q1, q2);
    const o3 = getOrientation(p2, q2, p1);
    const o4 = getOrientation(p2, q2, q1);

    // General case
    if (o1 !== o2 && o3 !== o4) {
        return true;
    }

    // Special Cases
    // p1, q1 and p2 are colinear and p2 lies on segment p1q1
    if (o1 === Orientation.Collinear && liesOnSegment(p1, p2, q1)) {
        return true;
    }

    // p1, q1 and p2 are colinear and q2 lies on segment p1q1
    if (o2 === Orientation.Collinear && liesOnSegment(p1, q2, q1)) {
        return true;
    }

    // p2, q2 and p1 are colinear and p1 lies on segment p2q2
    if (o3 === Orientation.Collinear && liesOnSegment(p2, p1, q2)) {
        return true;
    }

    // p2, q2 and q1 are colinear and q1 lies on segment p2q2
    if (o4 === Orientation.Collinear && liesOnSegment(p2, q1, q2)) {
        return true;
    }

    // Doesn't fall in any of the above cases
    return false;
}

const enum Orientation {
    Collinear,
    Clockwise,
    CounterClockwise
}

function getOrientation(p: Point, q: Point, r: Point): Orientation {
    const val = (q.y - p.y) * (r.x - q.x) - (q.x - p.x) * (r.y - q.y);

    if (isAlmostZero(val)) {
        return Orientation.Collinear;
    }

    return val > 0 ? Orientation.Clockwise : Orientation.CounterClockwise;
}

/**
 * Given three colinear points p, q, r, checks if point q lies on line segment 'p-r'
 */
function liesOnSegment(p: Point, q: Point, r: Point) {
    return (
        q.x <= Math.max(p.x, r.x) &&
        q.x >= Math.min(p.x, r.x) &&
        q.y <= Math.max(p.y, r.y) &&
        q.y >= Math.min(p.y, r.y)
    );
}
