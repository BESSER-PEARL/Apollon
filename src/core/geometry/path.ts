import { Delta } from "./delta";
import { areAlmostEqual } from "./math/floatingPoint";
import { Point, pointsAreEqual } from "./point";

export function beautifyPath(path: Point[]): Point[] {
    if (path.length <= 1) {
        return path;
    }

    path = removeConsecutiveIdenticalPoints(path);
    path = mergeConsecutiveSameAxisDeltas(path);
    path = flattenWaves(path);
    path = removeTransitNodes(path);

    return path;
}

function removeConsecutiveIdenticalPoints(path: Point[]): Point[] {
    const newPath: Point[] = [];
    for (const point of path) {
        const previousPoint = newPath[newPath.length - 1];
        if (!previousPoint || !pointsAreEqual(point, previousPoint)) {
            newPath.push(point);
        }
    }
    return newPath;
}

function removeTransitNodes(path: Point[]): Point[] {
    for (let i = 0; i < path.length - 2; i++) {
        const p = path[i];
        const q = path[i + 1];
        const r = path[i + 2];

        if (isHorizontalLineSegment(p, q, r) || isVerticalLineSegment(p, q, r)) {
            const pointsBeforeQ = path.slice(0, i + 1);
            const pointsAfterQ = path.slice(i + 2);
            const pathWithoutQ = [...pointsBeforeQ, ...pointsAfterQ];

            return removeTransitNodes(pathWithoutQ);
        }
    }

    return path;
}

function isHorizontalLineSegment(p: Point, q: Point, r: Point) {
    return (
        areAlmostEqual(p.y, q.y) &&
        areAlmostEqual(q.y, r.y) &&
        ((p.x >= q.x && q.x >= r.x) || (p.x <= q.x && q.x <= r.x))
    );
}

function isVerticalLineSegment(p: Point, q: Point, r: Point) {
    return (
        areAlmostEqual(p.x, q.x) &&
        areAlmostEqual(q.x, r.x) &&
        ((p.y <= q.y && q.y <= r.y) || (p.y >= q.y && q.y >= r.y))
    );
}

function mergeConsecutiveSameAxisDeltas(path: Point[]): Point[] {
    const deltas = computePathDeltas(path);

    if (deltas.length <= 1) {
        return path;
    }

    const newDeltas: Delta[] = [];

    for (const delta of deltas) {
        const previousDelta = newDeltas[newDeltas.length - 1];

        if (!previousDelta) {
            newDeltas.push(delta);
        } else if (
            (previousDelta.dx === 0 && delta.dx === 0) ||
            (previousDelta.dy === 0 && delta.dy === 0)
        ) {
            newDeltas[newDeltas.length - 1] = {
                dx: previousDelta.dx + delta.dx,
                dy: previousDelta.dy + delta.dy
            };
        } else {
            newDeltas.push(delta);
        }
    }

    return createPathFromDeltas(path[0], newDeltas);
}

function computePathDeltas(path: Point[]): Delta[] {
    const deltas: Delta[] = [];

    for (let i = 0; i < path.length - 1; i++) {
        const p = path[i];
        const q = path[i + 1];

        const dx = q.x - p.x;
        const dy = q.y - p.y;

        deltas.push({ dx, dy });
    }

    return deltas;
}

function createPathFromDeltas(start: Point, deltas: Delta[]): Point[] {
    const points = [start];
    let current = start;

    for (const { dx, dy } of deltas) {
        const x = current.x + dx;
        const y = current.y + dy;
        current = { x, y };
        points.push(current);
    }

    return points;
}

/**
 * Simplifies W-shaped path segments.
 *
 * e.g.
 *
 * O----O                0---------0
 *      |                          |
 *      O----O    ==>              0
 *           |                     |
 *           0                     0
 */
function flattenWaves(path: Point[]): Point[] {
    if (path.length < 4) {
        return path;
    }

    const deltas = computePathDeltas(path);
    const simplifiedDeltas = simplifyDeltas(deltas);

    const start = path[0];
    const simplifiedPath = createPathFromDeltas(start, simplifiedDeltas);

    return simplifiedPath;
}

function simplifyDeltas(deltas: Delta[]): Delta[] {
    for (let i = 0; i < deltas.length - 3; i++) {
        const d1 = deltas[i];
        const d2 = deltas[i + 1];
        const d3 = deltas[i + 2];
        const d4 = deltas[i + 3];

        if (
            d1.dy === 0 &&
            d2.dx === 0 &&
            d3.dy === 0 &&
            Math.sign(d1.dx) === Math.sign(d3.dx) &&
            Math.sign(d2.dy) === Math.sign(d4.dy)
        ) {
            return simplifyDeltas([
                ...deltas.slice(0, i),
                { dx: d1.dx + d3.dx, dy: 0 },
                { dx: 0, dy: d2.dy },
                ...deltas.slice(i + 3)
            ]);
        }

        if (
            d1.dx === 0 &&
            d2.dy === 0 &&
            d3.dx === 0 &&
            Math.sign(d1.dy) === Math.sign(d3.dy) &&
            Math.sign(d2.dx) === Math.sign(d4.dx)
        ) {
            return simplifyDeltas([
                ...deltas.slice(0, i),
                { dx: 0, dy: d1.dy + d3.dy },
                { dx: d2.dx, dy: 0 },
                ...deltas.slice(i + 3)
            ]);
        }
    }

    return deltas;
}
