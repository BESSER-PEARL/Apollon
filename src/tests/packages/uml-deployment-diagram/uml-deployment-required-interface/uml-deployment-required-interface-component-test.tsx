import * as React from 'react';
import { render } from '@testing-library/react';
import { Point } from "../../../../main/utils/geometry/point";
import { UMLDeploymentInterfaceRequired } from "../../../../main/packages/uml-deployment-diagram/uml-deployment-interface-required/uml-deployment-interface-required";
import { UMLInterfaceRequiredComponent } from "../../../../main/packages/common/uml-interface-required/uml-interface-required-component";

it('render the uml-deplyoment-required-interface-component', () => {
  const umlDeploymentInterfaceRequired: UMLDeploymentInterfaceRequired = new UMLDeploymentInterfaceRequired({ path: [new Point(0, 0), new Point(100, 100)] });
  // TODO: add store
  const { getByText } = render(
    <svg>
      <UMLInterfaceRequiredComponent element={umlDeploymentInterfaceRequired} />
    </svg>,
  );
  // TODO: expect
  // expect(getByText(umlDeploymentInterfaceRequired.name)).toBeInTheDocument();
});
