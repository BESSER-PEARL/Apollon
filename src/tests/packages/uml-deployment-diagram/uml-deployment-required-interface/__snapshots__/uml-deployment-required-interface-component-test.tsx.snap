// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the uml-deplyoment-required-interface-component 1`] = `
<body>
  <div>
    <svg>
      <g>
        <marker
          id="marker-f19e4479-d393-4239-bfdd-e2022b0853af"
          markerHeight="26"
          markerUnits="strokeWidth"
          markerWidth="26"
          orient="auto"
          refX="0"
          refY="0"
          stroke-dasharray="1,0"
          viewBox="0 0 26 26"
        >
          <path
            d="M 13 -13 a 13,13 0 0 0 0,26"
            fill="none"
            stroke="black"
            stroke-width="2"
          />
        </marker>
        <polyline
          fill="none"
          marker-end="url(#marker-f19e4479-d393-4239-bfdd-e2022b0853af)"
          points="0 0,100 97"
          stroke="black"
          stroke-width="1"
        />
      </g>
    </svg>
  </div>
</body>
`;
